Library/Homebrew/formula.rb

```diff
@@ -476,6 +476,29 @@ def compiler_failures
     active_spec.compiler_failures
   end
 
+  # The best spec we can upgrade the {Formula} to so it can satisfy dependency.
+  def best_upgrade_spec_sym
```

Comment:
> The main potentially ambiguous case I can think of is what to do when the HEAD spec has subsequently been removed from the formula.

Supposing the current implementation we have `HEAD` resolved to `HEAD` and if we don't have `HEAD` anymore, then `Formulary.factory` knows how to handle it and resolve everything properly even with `:head` spec passed.

> pin-like functionality for forcing upgrades to a specific formula to track HEAD.
@alyssais I think if a user wants to use these options then he's pretty much aware what's going on with his Cellar. So in my opinion it's better to satisfy whatever spec of dependencies he wants and then install some formula.